### PR TITLE
Implement PURGE to remove DVs from Delta tables

### PR DESCRIPTION
**Pull request by xupefei**
## Description

This PR introduces a `REORG TABLE ... APPLY (PURGE)` SQL command that can materialize soft-delete operations by DVs.

The command works by rewriting and bin-packing (if applicable) only files that have DVs attached, which is different from the `OPTIMIZE` command where all files (with and without) DV will be bin-packed. To achieve this, we hack the `OPTIMIZE` logic so files of any size with DVs will be rewritten.

Follow-up:
- Set the correct commit info. Now the resulting version is marked as `optimize` rather than `purge`.
- Clean up DVs from the filesystem.

## How was this patch tested?

New tests.

### CODE REVIEW

File: core/src/main/antlr4/io/delta/sql/parser/DeltaSqlBase.g4 [hunk 1]
```diff
@@ -87,8 +87,11 @@ statement
     | ALTER TABLE table=qualifiedName
         DROP CONSTRAINT (IF EXISTS)? name=identifier                    #dropTableConstraint
     | OPTIMIZE (path=STRING | table=qualifiedName)
-        (WHERE partitionPredicate = predicateToken)?
+        (WHERE partitionPredicate=predicateToken)?
```
Comment (larsk-db) [inline core/src/main/antlr4/io/delta/sql/parser/DeltaSqlBase.g4:90]: Is this change intended?

File: core/src/main/antlr4/io/delta/sql/parser/DeltaSqlBase.g4 [hunk 2]
```diff
@@ -87,8 +87,11 @@ statement
...
         (zorderSpec)?                                                   #optimizeTable
+    | REORG TABLE table=qualifiedName
```
Comment (larsk-db) [inline core/src/main/antlr4/io/delta/sql/parser/DeltaSqlBase.g4:92]: what about `path`? If it's legal for OPTIMIZE, why not legal for REORG TABLE?

File: core/src/main/antlr4/io/delta/sql/parser/DeltaSqlBase.g4 (hunk 1 above)
Comment (xupefei) [reply core/src/main/antlr4/io/delta/sql/parser/DeltaSqlBase.g4:90]: Yes, because I see other rules do not have spaces around `=`.

File: core/src/main/antlr4/io/delta/sql/parser/DeltaSqlBase.g4 (hunk 2 above)
Comment (xupefei) [reply core/src/main/antlr4/io/delta/sql/parser/DeltaSqlBase.g4:92]: Talked with Bart. He believes that the string literal version of OPTIMIZE is nonstandard, and is not something we would not recommend to users.

File: core/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala [hunk 3]
```diff
@@ -346,7 +346,28 @@ class DeltaSqlAstBuilder extends DeltaSqlBaseBaseVisitor[AnyRef] {
...
-      Option(ctx.partitionPredicate).map(extractRawText(_)).toSeq, Map.empty)(interleaveBy)
+      Option(ctx.partitionPredicate).map(extractRawText(_)).toSeq,
+      Map.empty)(interleaveBy)
+  }
+
+  /**
+   * Creates a [[ReorgTable]] logical plan.
+   * Examples:
+   * {{{
+   *   -- Physically delete dropped columns of target table
```
Comment (vkorukanti) [inline core/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:357]: dropped columns? -> dropped rows?

File: core/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala [hunk 4]
```diff
@@ -346,7 +346,28 @@ class DeltaSqlAstBuilder extends DeltaSqlBaseBaseVisitor[AnyRef] {
...
+   *   REORG TABLE (delta.`/path/to/table` | delta_table_name)
+   *    [WHERE partition_predicate] APPLY (PURGE)
+   * }}}
+   */
+  override def visitReorgTable(ctx: ReorgTableContext): AnyRef = withOrigin(ctx) {
+    if (ctx.table == null) {
+      throw new ParseException("REORG command requires a file path or table name.", ctx)
+    }
+
+    val targetIdentifier = visitMultipartIdentifier(ctx.table)
```
Comment (vkorukanti) [inline core/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:367]: why not use the `visitTableIdentifier`?

File: core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala [hunk 5]
```diff
@@ -164,18 +167,22 @@ class OptimizeExecutor(
...
   def optimize(): Seq[Row] = {
     recordDeltaOperation(txn.deltaLog, "delta.optimize") {
-      val minFileSize = sparkSession.sessionState.conf.getConf(
-        DeltaSQLConf.DELTA_OPTIMIZE_MIN_FILE_SIZE)
       val maxFileSize = sparkSession.sessionState.conf.getConf(
         DeltaSQLConf.DELTA_OPTIMIZE_MAX_FILE_SIZE)
-      require(minFileSize > 0, "minFileSize must be > 0")
       require(maxFileSize > 0, "maxFileSize must be > 0")
-
+      val (minFileSize, maxDeletedRowsRatio) = if (isPurge) {
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:173]: How about create a OptimizeConf and use it to pass the minFileSize and other required info? Can be done as part of the followup PRs.

File: core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala [hunk 6]
```diff
@@ -164,18 +167,22 @@ class OptimizeExecutor(
...
+        (0L, 0d) // Only selects files with DV
+      } else {
+        val minFileSize = sparkSession.sessionState.conf.getConf(
+          DeltaSQLConf.DELTA_OPTIMIZE_MIN_FILE_SIZE)
+        val maxDeletedRowsRatio = sparkSession.sessionState.conf.getConf(
+          DeltaSQLConf.DELTA_OPTIMIZE_MAX_DELETED_ROWS_RATIO)
+        require(minFileSize > 0, "minFileSize must be > 0")
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:180]: I think this check is already done as part of the conf setting.

File: core/src/main/scala/org/apache/spark/sql/delta/commands/ReorgTableCommand.scala [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  override val otherCopyArgs: Seq[AnyRef] = predicates :: Nil
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val command = OptimizeTableCommand(
+      Option(target.path.toString),
+      target.catalogTable.map(_.identifier),
+      predicates,
+      options = Map.empty,
+      isPurge = true)(zOrderBy = Nil)
+    command.run(sparkSession)
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/commands/ReorgTableCommand.scala:46]: is the command output expected to be same as the optimize? are the columns same?

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        .range(0, 100, 1, numPartitions = 5)
+        .write
+        .format("delta")
+        .save(path)
+      sql(s"DELETE FROM delta.`$path` WHERE id IN (0, 99)")
+      assert(log.update().allFiles.filter(_.deletionVector != null).count() === 2)
+      executePurge(path)
+      val (addFiles, _) = getFileActionsInLastVersion(log)
+      assert(addFiles.forall(_.deletionVector === null))
+      checkAnswer(
```
Comment (vkorukanti) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala:51]: check the files are combined into one?

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        sql(s"SELECT * FROM delta.`$path`"),
+        (1 to 98).toDF())
+    }
+  }
+
+  testWithDVs("Purge DVs") {
```
Comment (vkorukanti) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala:57]: `DeletionVectorTestUitls.withTempDeltaTable`?

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+      // Second purge is a noop
+      val versionBefore = log.update().version
+      executePurge(path)
+      val versionAfter = log.update().version
+      assert(versionBefore === versionAfter)
+    }
+  }
+
+  test("Purge a non-DV table is a noop") {
```
Comment (vkorukanti) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala:86]: does this compact the files?

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        .range(0, 100, 1, numPartitions = 5)
+        .write
+        .format("delta")
+        .save(path)
+      val versionBefore = log.update().version
+      executePurge(path)
+      val versionAfter = log.update().version
+      assert(versionBefore === versionAfter)
+    }
+  }
```
Comment (vkorukanti) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala:100]: We may need one or two tests for the case where reorg is run on a set of partitions.

File: core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala [hunk 12]
```diff
@@ -164,18 +167,22 @@ class OptimizeExecutor(
...
+        (minFileSize, maxDeletedRowsRatio)
+      }
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:182]: Also does this select the files that don't have DV and gets rewritten? Do we want that?

**Review by vkorukanti — CHANGES_REQUESTED** [review]

File: core/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala (hunk 3 above)
Comment (xupefei) [reply core/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:357]: I changed this to "rows and columns". Columns can be dropped when using column mapping.

File: core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala (hunk 12 above)
Comment (xupefei) [reply core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala:182]: It didn't select files without DVs, because with `minFileSize` set to `0` all files are de-selected unless they contain DVs. Inside `pruneCandidateFileList()`:
```
    // Select files that are small or have too many deleted rows
    files.filter(
      addFile => addFile.size < minFileSize || shouldCompactBecauseOfDeletedRows(addFile))
```

File: core/src/main/scala/org/apache/spark/sql/delta/commands/ReorgTableCommand.scala (hunk 7 above)
Comment (xupefei) [reply core/src/main/scala/org/apache/spark/sql/delta/commands/ReorgTableCommand.scala:46]: Yes, they're the same. The return schema contains two columns, `path` and `metrics`.

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala (hunk 8 above)
Comment (xupefei) [reply core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala:51]: Done!

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala (hunk 9 above)
Comment (xupefei) [reply core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala:57]: Nice to know, thanks!

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala (hunk 10 above)
Comment (xupefei) [reply core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala:86]: No. Added an assertion.

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala (hunk 11 above)
Comment (xupefei) [reply core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala:100]: Done!

File: core/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala (hunk 4 above)
Comment (xupefei) [reply core/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:367]: It's mainly because `UnresolvedTable` accepts a multipart table name rather than a `TableIdentifier`. I figured out it's possible to assemble the multipart name  manually:
```
    val targetIdentifier = visitTableIdentifier(ctx.table)
    val tableNameParts = targetIdentifier.database.toSeq :+ targetIdentifier.table
```

Comment (xupefei) [reply core/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:367]: We use `UnresolvedTable` here so it can throw a nice analysis error when the target table is a temp view.

File: core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala (hunk 5 above)
Comment (xupefei) [reply core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala]: Done in this PR!

**Review by larsk-db — APPROVED** [review]
LGTM

**Review by vkorukanti — APPROVED** [review]
